File: 0212-word-search-ii/solution.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isDone = False
        self.word = None
class Trie:
    def __init__(self):
        self.root = TrieNode()
    
    def add(self, word):
        cur = self.root
        for x in word:
            if x not in cur.children:
                cur.children[x] = TrieNode()
            cur = cur.children[x]
        cur.isDone = True
        cur.word = word


class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        res = []
        trie = Trie()
        
        # 1. Build the Trie
        for word in words:
            trie.add(word) # FIX: Call add on the instance 'trie'

        # 2. Define the DFS Helper (The Core Logic)
        def dfs(r, c, node):
            # TODO: Base Case 1 - Bounds Check
            # If r or c is out of bounds (less than 0 or >= len), return.
            if r < 0 or r >= len(board) or c < 0 or c >= len(board[0]):
                return
            
            # TODO: Base Case 2 - Visited Check
            # If board[r][c] is '#', it means we are already using this cell in the current path. Return.
            
            char = board[r][c]
            if char == '#':
                return 
            # TODO: Base Case 3 - Trie Path Check
            # If 'char' is NOT in 'node.children', we can't form a word here. Return.
            if char not in node.children:
                return
            # --- If we pass the checks, we are on a valid path ---
            
            # Move our Trie pointer forward
            next_node = node.children[char]
            
            # Check if we found a word
            if next_node.isDone:
                res.append(next_node.word)
                next_node.isDone = False
                # OPTIONAL: Set next_node.isDone = False to avoid adding the same word twice
            
            # Mark current cell as visited
            board[r][c] = '#' 
            
            # TODO: Recursive Step
            # Call dfs(..., ..., next_node) for all 4 directions:
            # (r+1, c), (r-1, c), (r, c+1), (r, c-1)
            dfs(r+1,c, next_node)
            dfs(r-1,c, next_node)
            dfs(r,c+1, next_node)
            dfs(r,c-1, next_node)
            # Backtrack: Restore the cell letter so other searches can use it
            board[r][c] = char

        # 3. Iterate over every cell to trigger the search
        for r in range(len(board)):
            for c in range(len(board[0])):
                # Optimization: Only start DFS if the cell contains a letter that starts a word
                if board[r][c] in trie.root.children:
                    dfs(r, c, trie.root)
                    
        return res

File: 0212-word-search-ii/solution.cpp
class TrieNode{
    public:
        vector<TrieNode*> nei;
        bool isWord;
        int wordIndex;
        TrieNode(){
            nei = vector<TrieNode*>(26, nullptr);
            isWord = false;
            wordIndex = 0;
        }
};
class Trie{
    public:
        TrieNode* root;
        Trie(){
            root = new TrieNode();
        }
        void addWord(string word, int index){
            TrieNode* node = root;
            for(char& cur : word){
                int i = cur - 'a';
                if(node->nei[i] == nullptr){
                    node->nei[i] = new TrieNode();
                }
                node = node->nei[i];
            }
            node->isWord = true;
            node->wordIndex = index;
        }
};
class Solution {
public:
    int ROW;
    int COL;
    void dfs(int r, int c, TrieNode* node, vector<string>& res, vector<vector<char>>& board, vector<string>& words){
        if(node->isWord == true){
            res.push_back(words[node->wordIndex]);
            node->isWord = false;
        }
        char temp = board[r][c];
        board[r][c] = '#';
        static const vector<vector<int>> directions = {{0, 1}, {1, 0}, {-1, 0}, {0, -1}};
        for(auto& dire : directions){
            int nr = r + dire[0];
            int nc = c + dire[1];
            if(nr >= ROW or nc >= COL or nr < 0 or nc < 0 or board[nr][nc] == '#'){
                continue;
            }

            if(node->nei[board[nr][nc] - 'a'] != nullptr){
                dfs(nr, nc, node->nei[board[nr][nc] - 'a'], res, board, words);
            }
        }
        board[r][c] = temp;
    }
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        vector<string> res;
        ROW = board.size();
        COL = board[0].size();

        Trie* trie = new Trie();

        for(int i = 0; i < words.size(); ++i){
            trie->addWord(words[i], i);
        }
        for(int r = 0; r < ROW; ++r){
            for(int c = 0; c < COL; ++c){
                int index = board[r][c] - 'a';
                if(trie->root->nei[index] != nullptr){
                    dfs(r, c, trie->root->nei[index], res, board, words);
                }
            }
        }
        return res;

    }
};
